etk-asm/src/bin/eas.rs

```diff
@@ -38,15 +43,41 @@ fn main() {
 fn run() -> Result<(), Error> {
     let opt: Opt = clap::Parser::parse();
 
+    let mut out_file_exists = false;
+    let mut out_file_content = vec![];
+    if let Some(o) = &opt.out {
+        if Path::new(o).exists() {
+            out_file_exists = true;
+            match std::fs::read(o) {
+                Ok(content) => out_file_content = content,
+                Err(e) => panic!("couldn't backup existing file: {}", e),
+            }
+        }
+    }
+
     let mut out: Box<dyn Write> = match opt.out {
-        Some(o) => Box::new(create(o)),
+        Some(ref o) => Box::new(create(o.to_path_buf())),
         None => Box::new(std::io::stdout()),
     };
 
     let hex_out = HexWrite::new(&mut out);
 
     let mut ingest = Ingest::new(hex_out);
-    ingest.ingest_file(opt.input)?;
+
+    if let Err(e) = ingest.ingest_file(opt.input) {
+        if out_file_exists {
+            match std::fs::write(&opt.out.unwrap(), &out_file_content) {
+                Ok(_) => (),
+                Err(e) => panic!("couldn't restore existing file.\n{}", e),
+            };
```

Comment:
The idea is to only modify the output if the compilation succeeds, but leave the file untouched otherwise?

This particular approach will fail to restore the file if, say, there's a panic.

What would you think of using something like [`tempfile`](https://crates.io/crates/tempfile) to make a temporary file in the same directory as the intended output, then moving the file into the intended output?

So to write to a file `a/b/test.foo`, you'd do:

1. Create `a/b/test.foo.tmp` using `tempfile`.
2. `ingest_file(...)`.
3. `std::fs::rename` from `a/b/test.foo.tmp` to `a/b/test.foo`.

Reply:
Hi, @SamWilsn. Yes, that was the idea - don’t create a file or overwrite the existing one with the same name if `eas` fails.

`tempfile` sounds like a good option.

Sorry I won’t be able to make the changes - I won’t have a computer with  me for a while.

I’d recommend the following approach:

1. Create a temporary file in the current directory
2. Ingest
3. If successful, create directories on the out path that don’t exist, and move + rename the file
4. If not, just remove the temporary file

This way, neither new directories nor an output file will be created if the compilation fails, which is cleaner.

Let me know what you think.